### DISCUSSION TITLE
Why need to use <link rel="preload" as="script"/> and <script async="" /> when building static html

### DISCUSSION THREAD

**Discussion by quynhbui-teqnological** [vercel/next.js#16154] [Help]
When I build static html to upload `out` folder to S3. I see in html file which generated include script tags and link tags like below:
![image](https://user-images.githubusercontent.com/52997226/90108484-93923980-dd74-11ea-8f0a-e258c8605e7f.png)
![image](https://user-images.githubusercontent.com/52997226/90108558-b3296200-dd74-11ea-9b85-22127b8188c4.png)

I don't know why we need to use `<link rel="preload" as="script"/>` tags and `<script async="" />`  tags in html file?When I try  to remove `<link rel="preload" as="script"/> `tags, the page stills work normal.
What happen if we only use `<script async="" />` to embed and load this files?

**Comment by deadcoder0904** [comment]
I think preload gives the priority to the files which are declared `rel="preload"` to load very early as the page loads up.

Same reason people use `font` as `preload` → https://github.com/tailwindlabs/blog.tailwindcss.com/blob/master/src/pages/_document.js#L41 as well as add it separately as `font-face` → https://github.com/tailwindlabs/blog.tailwindcss.com/blob/e69fad02113dbe400d772732109ae54890bf69e5/tailwind.config.js#L104

This way it loads fast. I'm pretty sure Next team does some optimizations for us that we don't even have to think about or worry, it's magic ✨

**Reply by quynhbui-teqnological** [reply]
I think when using `<link rel="preload" href="/_next/static/chunks/commons.1604956906f30c8c39d8.js" as="script">` file `commons.1604956906f30c8c39d8.js` will be loaded before `<script src="/_next/static/chunks/commons.1604956906f30c8c39d8.js" async=""></script>` embed and execute. Therefore, this script will be executed , doesn't take time to download it again?

**Reply by deadcoder0904** [reply]
Yep, it does make sense :)

**Reply by deadcoder0904** [reply]
I was just reading [Next.js Handbook from Flavio Copes](https://flaviocopes.nyc3.digitaloceanspaces.com/nextjs-handbook/book.pdf) & it states on page 24-25 under **The app bundles**,

# The app bundles

When we viewed the page source, we saw a bunch of JavaScript files being loaded:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1" />
    <meta name="next-head-count" content="2" />
    <link rel="preload" href="/_next/static/development/pages/index.js?ts=1572863116051" as="script" />
    <link rel="preload" href="/_next/static/development/pages/_app.js?ts=1572863116051" as="script" />
    <link rel="preload" href="/_next/static/runtime/webpack.js?ts=1572863116051" as="script" />
    <link rel="preload" href="/_next/static/runtime/main.js?ts=1572863116051" as="script" />
  </head>

  <body>
    <div id="__next">
      <div>
        <h1>Home page</h1>
      </div>
    </div>

    <script src="/_next/static/development/dll/dll_01ec57fc9b90d43b98a8.js?ts=1572863116051"></script>
    <script id="__NEXT_DATA__" type="application/json">
      { "dataManager": "[]", "props": { "pageProps": {} }, "page": "/", "query": {}, "buildId": "development", "nextExport": true, "autoExport": true }
    </script>
    <script async="" data-next-page="/" src="/_next/static/development/pages/index.js?ts=1572863116051"></script>
    <script async="" data-next-page="/_app" src="/_next/static/development/pages/_app.js?ts=1572863116051"></script>
    <script src="/_next/static/runtime/webpack.js?ts=1572863116051" async=""></script>
    <script src="/_next/static/runtime/main.js?ts=1572863116051" async=""></script>
  </body>
</html>
```

We have 4 JavaScript files being declared to be preloaded in the `head`, using `rel="preload" as="script"` :

- `/_next/static/development/pages/index.js` (96 LOC)

- `/_next/static/development/pages/_app.js` (5900 LOC)

- `/_next/static/runtime/webpack.js` (939 LOC)

- `/_next/static/runtime/main.js` (12k LOC)

This tells the browser to start loading those files as soon as possible before the normal rendering flow starts. Without those, scripts would be loaded with an additional delay, and this improves the page loading performance.

Then those 4 files are loaded at the end of the `body`, along with `/_next/static/development/dll/dll_01ec57fc9b90d43b98a8.js` (31k LOC), and a JSON snippet that sets some defaults for the page data:

```html
<script id="__NEXT_DATA__" type="application/json">
   {
   "dataManager": "[]",
   "props": {
   "pageProps":
   {}
   },
   "page": "/",
   "query": {},
   "buildId": "development",
   "nextExport": true,
   "autoExport": true
   }
</script>
```

The 4 bundle files loaded are already implementing one feature called code splitting. The `index.js` file provides the code needed for the `index` component, which serves the `/` route, and if we had more pages we'd have more bundles for each page, which will then only be loaded if needed - to provide a more performant load time for the page.

**Reply by quynhbui-teqnological** [reply]
I got it, thanks @deadcoder0904

**Reply by euskadi31** [reply]
It is possible to disable this behavior?

Thx

**Comment by michrome** [comment]
Hi @quynhbui-teqnological and @euskadi31 . Not sure I recommend it (because you lose all your Next.js goodness) but it is possible to disable this behaviour with `unstable_runtimeJS: false`:

```js
// pages/[id].js
export const config = {
  unstable_runtimeJS: false,
};

function Page({ id }) {
  return <h1>Page {id}</h1>;
}

export async function getStaticPaths() {
  return {
    paths: [{ params: { id: "1" } }, { params: { id: "2" } }],
    fallback: false,
  };
}

export async function getStaticProps({ params }) {
  return { props: { id: params.id } };
}

export default Page;
```

Running `next export` will give you:

```html
<!-- out/1.html -->
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <meta name="next-head-count" content="2" />
    <noscript data-n-css=""></noscript>
  </head>
  <body>
    <div id="__next">
      <h1>
        Page
        <!-- -->1
      </h1>
    </div>
  </body>
</html>
```

**Reply by euskadi31** [reply]
Thank You @michrome !

I need to control what is in the preloads, currently there are more than 40 JS in preloads which is blocking the rendering of the page.
